cmd/sysinfo-xl.go

```diff
@@ -0,0 +1,235 @@
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"strconv"
+	"sync"
+
+	"github.com/minio/minio/cmd/logger"
+	xnet "github.com/minio/minio/pkg/net"
+)
+
+// loadSysInfoXLAll - load all format config from all input disks in parallel.
+func loadSysInfoXLAll(storageDisks []StorageAPI) ([]*sysInfoConfig, []error) {
+	// Initialize sync waitgroup.
+	var wg = &sync.WaitGroup{}
+
+	// Initialize list of errors.
+	var sErrs = make([]error, len(storageDisks))
+
+	// Initialize format configs.
+	var sysinfos = make([]*sysInfoConfig, len(storageDisks))
+
+	// Load format from each disk in parallel
+	for index, disk := range storageDisks {
+		if disk == nil {
+			sErrs[index] = errDiskNotFound
+			continue
+		}
+		wg.Add(1)
+		// Launch go-routine per disk.
+		go func(index int, disk StorageAPI) {
+			defer wg.Done()
+
+			sysinfo, lErr := loadSysInfoXL(disk)
+			if lErr != nil {
+				sErrs[index] = lErr
+				return
+			}
+			sysinfos[index] = sysinfo
+		}(index, disk)
+	}
+
+	// Wait for all go-routines to finish.
+	wg.Wait()
+
+	// Return all formats and nil
+	return sysinfos, sErrs
+}
+
+func saveSysInfoXL(disk StorageAPI, format interface{}) error {
+	// Marshal and write to disk.
+	formatBytes, err := json.Marshal(format)
+	if err != nil {
+		return err
+	}
+
+	// Purge any existing temporary file, okay to ignore errors here.
+	defer disk.DeleteFile(minioMetaBucket, sysInfoConfigFileTmp)
+
+	// Append file `format.json.tmp`.
+	if err = disk.AppendFile(minioMetaBucket, sysInfoConfigFileTmp, formatBytes); err != nil {
+		return err
+	}
+
+	// Rename file `format.json.tmp` --> `format.json`.
+	return disk.RenameFile(minioMetaBucket, sysInfoConfigFileTmp, minioMetaBucket, sysInfoConfigFile)
+}
+
+// loadSysInfoXL - loads sysinfo.json from disk.
+func loadSysInfoXL(disk StorageAPI) (sysinfo *sysInfoConfig, err error) {
+	buf, err := disk.ReadAll(minioMetaBucket, sysInfoConfigFile)
+	if err != nil {
+		// 'file not found' and 'volume not found' as
+		// same. 'volume not found' usually means its a fresh disk.
+		if err == errFileNotFound || err == errVolumeNotFound {
+			var vols []VolInfo
+			vols, err = disk.ListVols()
+			if err != nil {
+				return nil, err
+			}
+			if len(vols) > 1 || (len(vols) == 1 &&
+				vols[0].Name != minioMetaBucket &&
+				vols[0].Name != "lost+found") {
+				// 'format.json' not found, but weformatConfigFile
+				// found user data.
+				return nil, errCorruptedFormat
+			}
+			// No other data found, its a fresh disk.
+			return nil, errUnformattedDisk
+		}
+		return nil, err
+	}
+
+	// Try to decode format json into formatConfigV1 struct.
+	sysinfo = &sysInfoConfig{}
+	if err = json.Unmarshal(buf, sysinfo); err != nil {
+		return nil, err
+	}
+
+	// Success.
+	return sysinfo, nil
+}
+
+// Get backend XL sysinfo.json in quorum `sysinfo.json`.
+func getSysInfoXLInQuorum(sysinfoConfigs []*sysInfoConfig) (*sysInfoConfig, error) {
+	return sysinfoConfigs[0], nil
```

Comment:
sysinfoConfigs[0] can easily be nil here when sysinfo.json is not found in the disk, isn't it ?

Reply:
Added a logic to calculate the ReadQuorum, instead of returning the first instance